remote_shell_service.go

```diff
@@ -0,0 +1,222 @@
+package main
+
+import (
+	"bufio"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/subtle"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/gliderlabs/ssh"
+	"github.com/google/uuid"
+	"go.uber.org/zap"
+	"golang.org/x/term"
+)
+
+type Console struct {
+	stdInLock  sync.Mutex
+	stdInPipe  io.Writer
+	stdOutPipe io.Reader
+	stdErrPipe io.Reader
+
+	sessionLock    sync.Mutex
+	remoteSessions map[uuid.UUID]ssh.Session
+}
+
+func makeConsole(stdin io.Writer, stdout io.Reader, stderr io.Reader) Console {
+	return Console{
+		stdInPipe:      stdin,
+		stdOutPipe:     stdout,
+		stdErrPipe:     stderr,
+		remoteSessions: map[uuid.UUID]ssh.Session{},
+	}
+}
+
+func (c *Console) WriteToStdIn(p []byte) (n int, err error) {
+	c.stdInLock.Lock()
+	n, err = c.stdInPipe.Write(p)
+	c.stdInLock.Unlock()
+
+	return n, err
+}
+
+func (c *Console) RegisterSession(id uuid.UUID, session ssh.Session) {
+	c.sessionLock.Lock()
+	c.remoteSessions[id] = session
+	c.sessionLock.Unlock()
+}
+
+func (c *Console) UnregisterSession(id uuid.UUID) {
+	c.sessionLock.Lock()
+	delete(c.remoteSessions, id)
+	c.sessionLock.Unlock()
+}
+
+func passwordHandler(ctx ssh.Context, password string, logger *zap.Logger) bool {
+	expectedPassword := os.Getenv("RCON_PASSWORD")
+	if expectedPassword == "" {
+		expectedPassword = "minecraft"
+	}
+
+	lengthComp := subtle.ConstantTimeEq(int32(len(password)), int32(len(expectedPassword)))
+	contentComp := subtle.ConstantTimeCompare([]byte(password), []byte(expectedPassword))
+	isValid := lengthComp == 1 && contentComp == 1
+	if !isValid {
+		logger.Warn(fmt.Sprintf("Remote console session rejected (%s/%s)", ctx.User(), ctx.RemoteAddr().String()))
+	}
+	return isValid
+}
+
+func handleSession(session ssh.Session, console *Console, logger *zap.Logger) {
+	// Setup state for the console session
+	sessionId := uuid.New()
+	_, _, isTty := session.Pty()
+	logger.Info(fmt.Sprintf("Remote console session accepted (%s/%s) isTTY: %t", session.User(), session.RemoteAddr().String(), isTty))
+	console.RegisterSession(sessionId, session)
+
+	lines := make(chan string)
+	clientExit := make(chan struct{})
+
+	go func() {
+		terminal := term.NewTerminal(session, "")
+		for {
+			line, err := terminal.ReadLine()
+			if err != nil {
+				if err != io.EOF {
+					logger.Error(fmt.Sprintf("Unable to read line from session (%s/%s)", session.User(), session.RemoteAddr().String()), zap.Error(err))
+				}
+				close(clientExit)
+				return
+			}
+
+			lines <- line
+		}
+	}()
+
+	shouldRun := true
+	for shouldRun {
```

Comment:
Labeling the loop and then break'ing in the 2nd and 3rd case I think would be slightly more Go-like:

```suggestion
	LineLoop:
	for {
```

Reply:
Oh, that's a neat feature.